crates/matrix-sdk/src/encryption/futures.rs

```diff
@@ -0,0 +1,85 @@
+use std::{
+    future::{Future, IntoFuture},
+    io::Read,
+    pin::Pin,
+};
+
+use eyeball::{shared::Observable as SharedObservable, Subscriber};
+
+use crate::{Client, Result, TransmissionProgress};
+
+/// Future returned by [`Client::prepare_encrypted_file`].
+#[allow(missing_debug_implementations)]
+pub struct PrepareEncryptedFile<'a, R: ?Sized> {
+    client: &'a Client,
+    content_type: &'a mime::Mime,
+    reader: &'a mut R,
+    send_progress: SharedObservable<TransmissionProgress>,
+}
+
+impl<'a, R: ?Sized> PrepareEncryptedFile<'a, R> {
+    pub(crate) fn new(client: &'a Client, content_type: &'a mime::Mime, reader: &'a mut R) -> Self {
+        Self { client, content_type, reader, send_progress: Default::default() }
+    }
+
+    /// Replace the default `SharedObservable` used for tracking upload
+    /// progress.
+    ///
+    /// Note that any subscribers obtained from
+    /// [`subscribe_to_send_progress`][Self::subscribe_to_send_progress]
+    /// will be invalidated by this.
+    pub fn with_send_progress_observable(
+        mut self,
+        send_progress: SharedObservable<TransmissionProgress>,
+    ) -> Self {
+        self.send_progress = send_progress;
+        self
+    }
+
+    /// Get a subscriber to observe the progress of sending the request
+    /// body.
+    pub fn subscribe_to_send_progress(&self) -> Subscriber<TransmissionProgress> {
+        self.send_progress.subscribe()
+    }
+}
+
+impl<'a, R> IntoFuture for PrepareEncryptedFile<'a, R>
+where
+    R: Read + Send + ?Sized + 'a,
+{
+    type Output = Result<ruma::events::room::EncryptedFile>;
+    #[cfg(target_arch = "wasm32")]
+    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output> + 'a>>;
+    #[cfg(not(target_arch = "wasm32"))]
+    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output> + Send + 'a>>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let Self { client, content_type, reader, send_progress } = self;
+        Box::pin(async move {
+            let mut encryptor = matrix_sdk_base::crypto::AttachmentEncryptor::new(reader);
+
+            let mut buf = Vec::new();
+            encryptor.read_to_end(&mut buf)?;
```

Comment:
If we end up removing the HTTP trait and only support `reqwests`, can we also get rid of this `read_to_end()` and instead pass in the `Reader` to `reqwests`?

Reply:
No, the only way to not buffer the request body is to put it in a `futures::Stream`, a `Read`er doesn't work. Also the `Stream` thing seems to not be available on WASM (which is why progress tracking also doesn't work on WASM, I should actually `cfg` out those methods).

Reply:
Hmm, ok, but I think that we'll have to special case this eventually so we don't use `read_to_end()` at least on non-WASM. Perhaps add an TODO for now, or a GH issue.

Reply:
This is not just for encryption, right? We also need to change Ruma to allow sending media without buffering the whole content in memory. Created https://github.com/matrix-org/matrix-rust-sdk/issues/2106.

Reply:
That's correct. Thanks.